File: payloads/c/fd_exhaustion.c
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>

int main(void) {
    int count = 0;
    while (1) {
        int fd = open("/dev/null", O_RDONLY);
        if (fd < 0) {
            fprintf(stderr, "fd exhaustion hit after %d fds: %s\n", count,
                    errno == EMFILE ? "EMFILE" : "other error");
            return 1;
        }
        count++;
    }
    printf("ESCAPED\n");
    return 0;
}

File: payloads/javascript/fd_exhaustion.js
const fs = require("fs");

const fds = [];
for (let i = 0; i < 1048576; i++) {
  try {
    fds.push(fs.openSync("/dev/null", "r"));
  } catch (e) {
    console.log("fd exhaustion at i=" + i + ":", e.message);
    break;
  }
}
console.log("opened fds:", fds.length);
